feat: add trigonometric functions

relativityhd opened the pull request:
# Add trigonometric functions (atan2 etc.)

Adds the following functions for all Floats:

- [x] Hyperbolic Sine (`sinh`)
- [x] Hyperbolic Cosine (`cosh`)
- [x] Inverse Sine (`asin`)
- [x] Inverse Hyperbolic Sine (`asinh`)
- [x] Inverse Cosine (`acos`)
- [x] Inverse Hyperbolic Cosine (`acosh`)
- [x] Inverse Tangent (`atan`)
- [x] Inverse Hyperbolic Tangent (`atanh`)
- [x] 2-argument arctangent (`atan2`)
- [x] Degrees (`degrees`)
- [x] Radians (`radians`)

## Open Questions / Missing parts

- [x] LLVM MIR implementation
- [x] Non f64 / f32 correctness
- [x] Metal atomic / safe operations

### LLVM MIR implementation

I have setup a placeholder currently with code which I assumed to work commented out.
It seems that support for these functions needs to be added in the tracel-llvm repository
, but I have no clue where.

### Non f64/f32

At some point there is limited support for the "special" floats and for e.g. powf which I used as an example a conversion was needed to normal floats.
I am unsure at which points this is necessary and where I can find out whether these conversions are necessary.

### Metal safe operations

Since I used the existing sin, cos, tanh and powf functions as examples on how to add the other functions, i stumbled across the implementation for metal for tanh:

```rust
    fn compile_instruction_tanh_scalar<T: Component<Self>>(
        f: &mut std::fmt::Formatter<'_>,
        input: T,
    ) -> std::fmt::Result {
        write!(f, "safe_tanh_scalar({input})")
    }
```

I couldn't find anything about this "safe" version in the metal documentation, but I am clearly not an expert.
For which functions is a "safe" implementation needed and which are fine without?

## Validate your PR with burn.

It is important that you make sure that you don't introduce any bugs in burn. 

### Instructions

- [] Create a new branch or fork of the [burn repo](https://github.com/Tracel-AI/burn)
- [] Update the main [Cargo.toml](https://github.com/tracel-ai/burn/blob/40aa993fb5969351a006b560ec89da5119dc9721/Cargo.toml#L159=L161) with this PR hash.
- [] Fix any broken tests or compilation errors in burn.
- [] Submit a PR in burn with your fixes and link it here.

relativityhd commented:
`cargo xtask validate` just takes too long on my machine, aborted after 10 minutes of execution. I have run `cargo test -p cubecl-wgpu` and `cargo test -p cubecl-core` instead and they seem to work.

wingertge commented:
Not sure if it might be a good idea to pull all the trigonometry functions (including the existing ones) into a new enum for the IR, because that's a lot of added stuff, and it's now enough to warrant its own category I think. We should at some point also think about applying a similar separation to the compilers themselves, but that's a larger rework that wouldn't go into this PR.

I'm currently looking into the MLIR functions for that stuff - the bindings are auto-generated I believe, so if the functions don't exist, they might be in a different namespace or need a different way to handle them.

As for the float types, casting to float for unsupported float types (i.e. F16, BF16) is reasonable. There do appear to be double versions of the functions, so that will work natively.

relativityhd commented on crates/cubecl-cpu/src/compiler/visitor/operation/arithmetic.rs:
```diff
@@ -28,6 +28,117 @@ impl<'a> Visitor<'a> {
                 let result = self.append_operation_with_result(operation);
                 self.insert_variable(out, result);
             }
+            Arithmetic::Sinh(_sinh) => {
```
I implemented everything. Now I get this error message when running tests:
error: cannot be converted to LLVM IR: missing `LLVMTranslationDialectInterface` registration for dialect for op: math.atan2
However, according to code hints everything seems fine...

marcantoinem replied:
I tried your code and I get the same error the pass seems to do nothing. I didn't use it anywhere else in the MLIR compiler so it must be a problem on the C API of MLIR that is not running correctly this pass. For the moment you can use LLVM intrisic I think, it will needs more investigation to find why registering the math_to_llvm doesn't work. The only problem with using llvm intrisic instead of math is for portability if we want to port the MLIR compiler to GPU,. but it is not a concern right now.

relativityhd replied:
The arc-* operations are not available in the llvm intrinsic, hence I disabled them for now until the math module becomes available.

relativityhd replied:
Yes and as far as I can see there is nothing which would prevent acosh etc. from loading, since the dialects are directly imported from mlir:

https://github.com/tracel-ai/tracel-llvm/blob/9d51e26fab8542949c94f7c14c2519113648bf27/crates/tracel-mlir-sys/wrapper.h#L19

and

https://github.com/tracel-ai/tracel-llvm/blob/9d51e26fab8542949c94f7c14c2519113648bf27/crates/tracel-mlir-rs/src/pass/conversion.rs#L35

and

https://github.com/tracel-ai/tracel-llvm/blob/9d51e26fab8542949c94f7c14c2519113648bf27/crates/tracel-mlir-rs/src/dialect/ods.rs#L144

Maybe some files are missing in https://github.com/tracel-ai/tracel-llvm/blob/9d51e26fab8542949c94f7c14c2519113648bf27/crates/tracel-mlir-rs/src/dialect/ods.rs#L146 ?

relativityhd replied:
I have no clue how I could generate these files or where they are generated ...

relativityhd commented:
@wingertge 

> Not sure if it might be a good idea to pull all the trigonometry functions (including the existing ones) into a new enum for the IR, because that's a lot of added stuff, and it's now enough to warrant its own category I think. 
Must I implement this in this PR or will this be part of another refactor?

> As for the float types, casting to float for unsupported float types (i.e. F16, BF16) is reasonable. There do appear to be double versions of the functions, so that will work natively.
Can you please double-check my implementation? I don't really understand at which point the right functions are written to the shader for each dialect...

wingertge commented:
> @wingertge
> 
> > Not sure if it might be a good idea to pull all the trigonometry functions (including the existing ones) into a new enum for the IR, because that's a lot of added stuff, and it's now enough to warrant its own category I think.
> 
> Must I implement this in this PR or will this be part of another refactor?

I think that would be better to do after all the outstanding work has been merged, so I'll do it separately.
One more thing I'd add retroactively is that since there's no intrinsics for converting degrees/radians, I think it would be better to just implement those as `cube` functions in `cubecl-std`, rather than the compiler. The implementation would be trivial:
```rust
#[cube]
pub fn to_degrees<F: Float>(val: F) -> F {
    val * F::new(180.0 / f32::PI)
}
```
It would keep it in one place rather than implemented separately for each compiler.

wingertge commented:
> Can you please double-check my implementation? I don't really understand at which point the right functions are written to the shader for each dialect...

CUDA is already correct, not sure what the WGSL behaviour even is because it's an experimental extension. It might just support f16 overloads on those functions by default, but I can't yet test it because of some issues with features on Vulkan.
For SPIR-V I think it would be best implemented by a preprocessor but they're poorly documented, so I think the easiest way to do it is to open an issue to track that it still needs to be done and I'll deal with it myself.

relativityhd commented:
> there's no intrinsics for converting degrees/radians

There are intrinsics for WGSL and SPIR-V (at least rspirv-ext), also Rusts f32 and f64 do have to_degrees and to_radians operations. Also, as a user, I would probably search in the Float Namespace for these functions instead of the cubecl-std. That's why I think having them there would make still sense.

I removed the dialect specific compile_instruction_xxx_scalar calls and instead hardcode the operations directly for the cubecl-cpp, since all CUDA, HIP and Metal should convert implicit to the right type and share the same syntax.
I also added the to_degrees and to_radians functions to the cubecl-std, so that you can decide which one fits best / should survive a refactor.

relativityhd commented:
@nathanielsimard yes the CI is failing because as I mentioned this PR is blocked by the MLIR melior project, because the math module there wont register correctly, as @marcantoinem mentioned.
The tests are failing because of the empty implementations of the functions depending on the melior llvm math module.

> We may also remove some functions in the trigo modules, unsure they are necessary.

Sure, just tell which ones I should remove.

relativityhd commented:
I have added you all to my repo in case I am blocking your development, just in case

relativityhd commented:
I have merged from main, but still no success. Still got the `error: cannot be converted to LLVM IR: missing `LLVMTranslationDialectInterface` registration for dialect for op: func.func` when running `cargo xtask validate`...

I've tried running with the debug flag you mentioned `CUBECL_MLIR_DEBUG="./debug-mlir/" cargo xtask validate` but I can't see a new directory anywhere...

relativityhd commented:
OMG yes it finally works, thank you!

So I guess this PR is now ready for merge! @wingertge @nathanielsimard

relativityhd commented on crates/cubecl-cpp/src/shared/dialect.rs:
```diff
@@ -633,6 +633,10 @@ pub trait DialectInstructions<D: Dialect> {
         }
     }
 
+    fn compile_instruction_atan2(f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```
Removed it 👍🏻

antimora commented:
@nathanielsimard when is the plan to add these to Burn?
